osu.Game/Graphics/Containers/Markdown/OsuMarkdownFencedCodeBlock.cs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using Markdig.Syntax;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers.Markdown;
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Overlays;
+
+namespace osu.Game.Graphics.Containers.Markdown
+{
+    public class OsuMarkdownFencedCodeBlock : MarkdownFencedCodeBlock
+    {
+        private Box background;
+        private MarkdownTextFlowContainer textFlow;
+
+        public OsuMarkdownFencedCodeBlock(FencedCodeBlock fencedCodeBlock)
+            : base(fencedCodeBlock)
+        {
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OverlayColourProvider colourProvider)
+        {
+            // TODO : Change to monospace font to match with osu-web
+            background.Colour = colourProvider.Background6;
+            textFlow.Colour = colourProvider.Light1;
```

Comment:
At first I also thought about this. But when I try something like this for example.

```csharp
    public class OsuMarkdownFencedCodeBlock : MarkdownFencedCodeBlock
    {

        [Resolved]
        private OverlayColourProvider colourProvider { get; set; }

        public OsuMarkdownFencedCodeBlock(FencedCodeBlock fencedCodeBlock)
            : base(fencedCodeBlock)
        {
        }

        protected override Drawable CreateBackground()
        {
            return new Box
            {
                RelativeSizeAxes = Axes.Both,
                Colour = colourProvider.Background6,
            };
        }

        public override MarkdownTextFlowContainer CreateTextFlow()
        {
            var textFlow = base.CreateTextFlow();
            textFlow.Colour = colourProvider.Light1;
            return textFlow;
        }
    }

```

It just doesn't work and crash.

Am I missing something?

Reply:
You're right. This is an unfortunate consequence of how DI works. Because the *base* class's BDL is what is accessing `Create{Background,TextFlow}()`, rather than this one's, this class's `[Resolved]` fields won't have been injected yet at that stage, therefore leading to nullrefs.

This *could* be worked around by using `Dependencies.Get<OverlayColourProvider>()`, but I'm not sure I want to recommend that precedent... This feels like a yet-undiscovered hole in how we do injection, if I'm honest. Maybe a framework change that always injects all `[Resolved]` fields before running any BDLs is needed here? Would need further opinions on this from others.

Reply:
I agree all `[Resolved]` members should be injected before any BDL is invoked, doesn't sound to cause any harm.

Reply:
This could be fixed by making derived components, rather than weirdly assigning the component to a field.

Reply:
@peppy so basically we add private class as derived component inside this class like this

```csharp
    public class OsuMarkdownFencedCodeBlock : MarkdownFencedCodeBlock
    {
        public OsuMarkdownFencedCodeBlock(FencedCodeBlock fencedCodeBlock)
            : base(fencedCodeBlock)
        {
        }

        protected override Drawable CreateBackground() => new CodeBlockBackground();

        public override MarkdownTextFlowContainer CreateTextFlow() => new CodeBlockTextFlowContainer();

        private class CodeBlockBackground : Box
        {
            [BackgroundDependencyLoader]
            private void load(OverlayColourProvider colourProvider)
            {
                RelativeSizeAxes = Axes.Both;
                Colour = colourProvider.Background6;
            }
        }

        private class CodeBlockTextFlowContainer : OsuMarkdownTextFlowContainer
        {
            [BackgroundDependencyLoader]
            private void load(OverlayColourProvider colourProvider)
            {
                Colour = colourProvider.Light1;
                Margin = new MarginPadding(10);
            }
        }
```

or am I wrong?

Reply:
That's correct.

Reply:
The drawback of this pattern is I should re-apply all of customization, for example, from `base.CreateTextFlow()` into the custom derived component `CodeBlockTextFlowContainer`. But after a slightly check in o!f, mostly it just one or two customization.